### PR TITLE
Implement knapsack exercise

### PR DESCRIPTION
**Pull request by sonapraneeth-a**
Implementation of knapsack exercise.

Fixes [#1747](https://github.com/exercism/java/issues/1747)

---

Reviewer Resources:

[Track Policies](https://github.com/exercism/java/blob/master/POLICIES.md#event-checklist)

### CODE REVIEW

**Comment by sonapraneeth-a** [comment]
Not sure why ParallelLetterFrequencyTest is failing in [build](https://travis-ci.org/exercism/java/jobs/593468407) and passing in [build](https://travis-ci.org/exercism/java/builds/593421754) though I did not make any changes there.

**Comment by lemoncurry** [comment]
> Any idea on what to do further?

I think it has to do with an update today of gradle. We may need to reverse it or disable this test because it produces the following error:
`java.lang.OutOfMemoryError: Java heap space`

**Comment by FridaTveit** [comment]
That's unfortunate, I'm sorry it's impacted your PR @sonapraneeth-a! I'd vote for removing the parallel-letter-frequency test. It's quite a complicated test that could easily confuse users and it's not based on any canonical data. What do you think @lemoncurry? 🙂

**Comment by lemoncurry** [comment]
@FridaTveit I agree, it should be removed.

**Comment by FridaTveit** [comment]
I'll open a PR with it removed ASAP! 🙂

**Comment by FridaTveit** [comment]
PR opened: https://github.com/exercism/java/pull/1766 🙂

**Comment by sonapraneeth-a** [comment]
How do I re-run the travis build once your PR is merged @FridaTveit ?

**Comment by FridaTveit** [comment]
@sonapraneeth-a you'll need to rebase your changes on master so that your branch has my PR on it. Here are some resources for that if you're not familiar with rebasing:
- https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase
- https://git-scm.com/book/en/v2/Git-Branching-Rebasing 
🙂

**Comment by lemoncurry** [comment]
@sonapraneeth-a I restarted the job :smile: The check is now successful.

File: config.json [hunk 1]
```diff
@@ -1348,6 +1348,18 @@
...
+    {
+      "slug": "knapsack",
+      "uuid": "ac179b77-98a5-4daf-9773-3d68b6cd8548",
+      "core": false,
+      "unlocked_by": null,
+      "difficulty": 8,
+      "topics": [
+        "algorithms",
+        "arrays",
+        "control_flow_loops"
```
Comment (lemoncurry) [inline config.json]: This topic is not mentioned in the [topics list](https://github.com/exercism/problem-specifications/blob/master/TOPICS.txt).
Maybe there is another topic that might fit from the official list?

File: config.json [hunk 2]
```diff
@@ -1348,6 +1348,18 @@
         "math"
       ]
     },
+    {
+      "slug": "knapsack",
+      "uuid": "ac179b77-98a5-4daf-9773-3d68b6cd8548",
+      "core": false,
+      "unlocked_by": null,
```
Comment (lemoncurry) [inline config.json]: Could you please define an exercise that will unlock this one? e.g. saddle-points

File: exercises/knapsack/.meta/hints.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+- Use recursion
```
Comment (lemoncurry) [inline exercises/knapsack/.meta/hints.md]: Great hints :smile:

File: exercises/knapsack/.meta/src/reference/java/Knapsack.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
+import java.util.ArrayList;
+
+/**
+ * Knapsack item with a weight and value
+ */
```
Comment (lemoncurry) [inline exercises/knapsack/.meta/src/reference/java/Knapsack.java]: Please remove any comments

File: exercises/knapsack/.meta/src/reference/java/Knapsack.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    int weight;
+    // Value of the item
+    int value;
+
+    Item(int itemWeight, int itemValue) {
+        this.weight = itemWeight;
+        this.value = itemValue;
+    }
+    
+}
```
Comment (lemoncurry) [inline exercises/knapsack/.meta/src/reference/java/Knapsack.java]: Please extract this class

File: exercises/knapsack/.meta/src/reference/java/Knapsack.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+class Knapsack {
```
Comment (lemoncurry) [inline exercises/knapsack/.meta/src/reference/java/Knapsack.java]: Concerning the algorithm you used:
I think it is too complicated and hard to follow (maybe this will improve after removing comments).
I checked for other algorithms solving this well known coding problem, and, for example, I like [this article](https://medium.com/@ssaurel/solving-the-knapsack-problem-in-java-c985c71a7e64) and the solution presented.

File: config.json [hunk 7]
```diff
@@ -1348,6 +1348,18 @@
...
+      "difficulty": 8,
```
Comment (lemoncurry) [inline config.json]: I would actually say it is more like a difficulty 4-5 problem.

File: exercises/knapsack/src/main/java/Knapsack.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import java.util.ArrayList;
```
Comment (lemoncurry) [inline exercises/knapsack/src/main/java/Knapsack.java]: For this file, you may also use the [stub template](https://github.com/exercism/java/tree/master/_template/src/main/java), if it is an exercise with difficulty >4.

Comment (sonapraneeth-a) [reply exercises/knapsack/src/main/java/Knapsack.java]: Changing this file to use stub template, how would one know if the input is ```ArrayList``` or not?

File: exercises/knapsack/.meta/src/reference/java/Knapsack.java (hunk 6 above)
Comment (sonapraneeth-a) [reply exercises/knapsack/.meta/src/reference/java/Knapsack.java]: I've added comments so that it would easy to follow. I've coded three approaches

- Plain recursion
- Top down optimization
- Bottom up optimization

I'm not sure if it would be good, if I directly write the bottom up approach. If you feel the other way round, I'll remove all other code and keep the bottom up approach in the reference implementation.

File: exercises/knapsack/.meta/src/reference/java/Knapsack.java (hunk 5 above)
Comment (sonapraneeth-a) [reply exercises/knapsack/.meta/src/reference/java/Knapsack.java]: in a separate file?

File: config.json (hunk 1 above)
Comment (sonapraneeth-a) [reply config.json]: Recursion and dynamic programming would be suitable topics. But none of them are present in the file shared.

File: exercises/knapsack/.meta/src/reference/java/Knapsack.java (hunk 6 above)
Comment (lemoncurry) [reply exercises/knapsack/.meta/src/reference/java/Knapsack.java]: I see. But for the reference implementation, you only need to provide one algorithm that solves the problem. Which one would you use?

Comment (sonapraneeth-a) [reply exercises/knapsack/.meta/src/reference/java/Knapsack.java]: As Bottom up is the best one, I may use that if only one algorithm is supposed to be there.

File: exercises/knapsack/src/main/java/Knapsack.java (hunk 8 above)
Comment (lemoncurry) [reply exercises/knapsack/src/main/java/Knapsack.java]: You will know it from the test cases.

File: exercises/knapsack/.meta/src/reference/java/Knapsack.java (hunk 5 above)
Comment (lemoncurry) [reply exercises/knapsack/.meta/src/reference/java/Knapsack.java]: Yes, that would be great. We have many exercises and hence reference implementations that use more than one class.

File: exercises/knapsack/src/main/java/Knapsack.java (hunk 8 above)
Comment (sonapraneeth-a) [reply exercises/knapsack/src/main/java/Knapsack.java]: Done.

File: config.json (hunk 7 above)
Comment (sonapraneeth-a) [reply config.json]: Updated

File: exercises/knapsack/.meta/src/reference/java/Knapsack.java (hunk 6 above)
Comment (sonapraneeth-a) [reply exercises/knapsack/.meta/src/reference/java/Knapsack.java]: Cleaned the file

File: exercises/knapsack/.meta/src/reference/java/Knapsack.java (hunk 5 above)
Comment (sonapraneeth-a) [reply exercises/knapsack/.meta/src/reference/java/Knapsack.java]: Extracted Item into Item.java

File: exercises/knapsack/.meta/src/reference/java/Knapsack.java (hunk 4 above)
Comment (sonapraneeth-a) [reply exercises/knapsack/.meta/src/reference/java/Knapsack.java]: Removed all comments

File: config.json (hunk 2 above)
Comment (sonapraneeth-a) [reply config.json]: Currently using saddle-points to unlock knapsack

**Comment by sonapraneeth-a** [comment]
@lemoncurry , Please review. I've made necessary changes from my side.
